### PR TITLE
Aggregation by POST request to /tiles endpoint

### PR DESCRIPTION
**Pull request by keller-mark**
## Description

> What was changed in this pull request?

Adds support for making POST request to `/tiles` to get multivec tiles aggregated in specified groups of rows, according to the specified aggregation function.

```json
{
      "my_multivec_tileset_uid": {
        "agg_groups": [[0, 1], [2, 3, 4]],
        "agg_func": "mean"
    }
}
```

Supports caching by converting the JSON array to an md5 hash and appending the aggregation function name when setting/getting in the redis database. (like `my_multivec_tileset_uid.1.1000.fjdksaljfkdlasjfkldsajfkl.mean`)

> Why is it necessary?

Sehi’s performance testing results show that the majority of the bottlenecks come from the network https://github.com/sehilyi/higlass-horizontal-multivec-performance
If aggregation is done on the server, there is less information to send over the network.

Fixes #___

## Checklist

- [X] Unit tests added or updated
- [ ] Documentation added or updated
- [X] Updated CHANGELOG.md

### CODE REVIEW

File: tilesets/generate_tiles.py [hunk 1]
```diff
@@ -122,12 +122,23 @@ def generate_1d_tiles(filename, tile_ids, get_data_function):
...
     tile_list: [(tile_id, tile_data),...]
         A list of tile_id, tile_data tuples
     '''
+
+    agg_func_map = {
+        "sum": lambda x: np.sum(x, axis=0),
+        "mean": lambda x: np.mean(x, axis=0),
+        "median": lambda x: np.median(x, axis=0),
+        "std": lambda x: np.std(x, axis=0),
+        "var": lambda x: np.var(x, axis=0)
```
Comment (sehilyi) [inline tilesets/generate_tiles.py:139]: Do we need to support `max` and `min` functions as well? These can be useful when people are interested in seeing extreme values in groups.

File: tilesets/generate_tiles.py [hunk 2]
```diff
@@ -488,7 +505,7 @@ def generate_tiles(tileset_tile_ids):
     tile_list: [(tile_id, tile_data),...]
         A list of tile_id, tile_data tuples
     '''
-    tileset, tile_ids, raw = tileset_tile_ids
+    tileset, tile_ids, raw, agg_info = tileset_tile_ids
```
Comment (sehilyi) [inline tilesets/generate_tiles.py:508]: Could you update the comments on the above to include `agg_info` and `raw`?

File: tilesets/generate_tiles.py [hunk 3]
```diff
@@ -136,6 +149,12 @@ def generate_1d_tiles(filename, tile_ids, get_data_function):
 
         dense = get_data_function(filename, tile_position)
 
+        if agg_info != None:
+            agg_func_name = agg_info["agg_func"]
+            agg_group_arr = [ x if type(x) == list else [x] for x in agg_info["agg_groups"] ]
+            assert(agg_func_name in agg_func_map)
```
Comment (pkerpedjiev) [inline tilesets/generate_tiles.py:155]: Might be good to return some meaningful error message here. Maybe raise an exception and then catch it later and return as `{"error": message}` in the tile return value.

File: tilesets/views.py [hunk 4]
```diff
@@ -393,6 +394,23 @@ def tiles(request):
...
+            for tileset_id, agg_info in tileset_to_agg_info.items():
+                assert("agg_groups" in agg_info.keys())
+                assert("agg_func" in agg_info.keys())
+                agg_groups = agg_info["agg_groups"]
+                agg_func_name = agg_info["agg_func"]
+                agg_hash = '.' + hashlib.md5(json.dumps(agg_groups).encode('utf-8')).hexdigest() + '.' + agg_func_name
+                tileset_to_agg_info[tileset_id]["agg_hash"] = agg_hash
+        except:
+            return JsonResponse({
+                'error': 'Unable to parse request body as JSON'
```
Comment (pkerpedjiev) [inline tilesets/views.py:411]: This should be more meaningful.

**Comment by keller-mark** [comment]
Thanks for the feedback, I agree with all of the points you bring up. However, I think the options should be scoped to a tileSET rather than a tile, so that you don't need to duplicate the options when you request multiple tiles from a tileset in the same request.

And it seems like the convention for casing in the JSON is camelCase, correct?

How do you feel about the following?

```json
[
  {
    "tilesetUid": "x",
    "tiles": ["x.0.0", "x.0.1", "x.0.2"],
    "options": {
       "aggGroups": [[1, 2], [3, 4, 5]],
       "aggFunc": "sum",
    }
  }
]
```

**Comment by pkerpedjiev** [comment]
> However, I think the options should be scoped to a tileSET rather than a tile, so that you don't need to duplicate the options when you request multiple tiles from a tileset in the same request.

👍 Might be good to remove the tilesetUid from the tile ids so that it's not duplicated.

**Comment by keller-mark** [comment]
Thanks for sharing the pydantic package. I took a look but decided to use `jsonschema`  because I am more familiar with it and the `"oneOf"` expression feels natural for different types of options that may need to be supported in the future.

I added a 1,000 tile limit, does that number seem appropriate?

File: tilesets/views.py [hunk 5]
```diff
@@ -477,9 +526,13 @@ def tiles(request):
     tiles_to_return = {}
 
     for (tile_id, tile_value) in generated_tiles:
-
+        tileset_uuid = tgt.extract_tileset_uid(tile_id)
         try:
-            rdb.set(tile_id, pickle.dumps(tile_value))
+            if tileset_uuid in tileset_to_options:
+                tileset_options = tileset_to_options[tileset_uuid]
+                rdb.set(tile_id + tileset_options["options_hash"], pickle.dumps(tile_value))
```
Comment (pkerpedjiev) [inline tilesets/views.py:533]: Nice that you thought of this 👍

**Pull request merged**